db/badger_db.go

```diff
@@ -0,0 +1,323 @@
+	valueItem := new(badger.KVItem)
+	if err := b.kv.Get(key, valueItem); err != nil {
+		// Unfortunate that Get can't return errors
+		// TODO: Propose allowing DB's Get to return errors too.
+		common.PanicCrisis(err)
+	}
+	var valueSave []byte
+	err := valueItem.Value(func(origValue []byte) error {
+		// TODO: Decide if we should just assign valueSave to origValue
+		// since here we aren't dealing with iterators directly.
+		valueSave = make([]byte, len(origValue))
+		copy(valueSave, origValue)
+		return nil
+	})
+	if err != nil {
+		// TODO: ditto:: Propose allowing DB's Get to return errors too.
+		common.PanicCrisis(err)
+	}
+	return valueSave
+}
+
+// noUserMeta signifies blank metadata for a kv entry
+// since badger.Set accepts an optional userMeta field
+var noUserMeta byte = 0x00
+
+func panicOnErr(err error) {
+	if err != nil {
+		common.PanicCrisis(err)
+	}
+}
+
+func (b *BadgerDB) Set(key, value []byte) {
+	b.kv.SetAsync(key, value, noUserMeta, panicOnErr)
+}
+
+func (b *BadgerDB) SetSync(key, value []byte) {
+	if err := b.kv.Set(key, value, noUserMeta); err != nil {
+		common.PanicCrisis(err)
+	}
+}
+
+func (b *BadgerDB) Delete(key []byte) {
+	b.kv.DeleteAsync(key, panicOnErr)
+}
+
+func (b *BadgerDB) DeleteSync(key []byte) {
+	if err := b.kv.Delete(key); err != nil {
+		common.PanicCrisis(err)
+	}
+}
+
+func (b *BadgerDB) Close() {
+	if err := b.kv.Close(); err != nil {
+		common.PanicCrisis(err)
+	}
+}
+
+func (b *BadgerDB) Fprint(w io.Writer) {
+	bIter := b.Iterator().(*badgerDBIterator)
+	defer bIter.Close()
+
+	var bw *bufio.Writer
+	if bbw, ok := w.(*bufio.Writer); ok {
+		bw = bbw
+	} else {
+		bw = bufio.NewWriter(w)
+	}
+	defer bw.Flush()
+
+	i := uint64(0)
+	for bIter.rewind(); bIter.valid(); bIter.Next() {
+		k, v := bIter.kv()
+		fmt.Fprintf(bw, "[%X]:\t[%X]\n", k, v)
+		i += 1
+		if i%1024 == 0 {
+			bw.Flush()
+			i = 0
+		}
+	}
+}
+
+func (b *BadgerDB) Print() {
+	bw := bufio.NewWriter(os.Stdout)
+	b.Fprint(bw)
+}
+
+func (b *BadgerDB) Iterator() Iterator {
+	dbIter := b.kv.NewIterator(badger.IteratorOptions{
+		PrefetchValues: true,
+
+		// Arbitrary PrefetchSize
+		PrefetchSize: 10,
+	})
+	// Ensure that we are always at the zeroth item
+	dbIter.Rewind()
+	return &badgerDBIterator{iter: dbIter}
+}
+
+func (b *BadgerDB) Stats() map[string]string {
+	return nil
+}
+
+func (b *BadgerDB) NewBatch() Batch {
+	return &badgerDBBatch{db: b}
+}
+
+var _ Batch = (*badgerDBBatch)(nil)
+
+type badgerDBBatch struct {
+	entriesMu sync.Mutex
+	entries   []*badger.Entry
+
+	db *BadgerDB
+}
+
+func (bb *badgerDBBatch) Set(key, value []byte) {
+	bb.entriesMu.Lock()
+	bb.entries = append(bb.entries, &badger.Entry{
+		Key:   key,
+		Value: value,
+	})
+	bb.entriesMu.Unlock()
+}
+
+// Unfortunately Badger doesn't have a batch delete
+// The closest that we can do is do a delete from the DB.
+// Hesitant to do DeleteAsync because that changes the
+// expected ordering
+func (bb *badgerDBBatch) Delete(key []byte) {
+	bb.db.Delete(key)
```

Comment:
no error handling?

Reply:
The 'error handling' present with our structure doesn't return an error yet I am apprehensive about adding panics in Delete.